python/ray/train/predictor.py

```diff
@@ -139,35 +166,58 @@ def predict(self, data: DataBatchType, **kwargs) -> DataBatchType:
         Args:
             data: A batch of input data of type ``DataBatchType``.
             kwargs: Arguments specific to predictor implementations. These are passed
-            directly to ``_predict_pandas``.
+            directly to ``_predict_numpy`` or ``_predict_pandas``.
 
         Returns:
             DataBatchType:
                 Prediction result. The return type will be the same as the input type.
         """
-        data_df = convert_batch_type_to_pandas(data, self._cast_tensor_columns)
-
         if not hasattr(self, "_preprocessor"):
             raise NotImplementedError(
                 "Subclasses of Predictor must call Predictor.__init__(preprocessor)."
             )
-
         if self._preprocessor:
-            data_df = self._preprocessor.transform_batch(data_df)
+            data = self._preprocessor.transform_batch(data)
+        try:
+            batch_format = TYPE_TO_ENUM[type(data)]
+        except KeyError:
+            raise RuntimeError(
+                f"Invalid input data type of {type(data)}, supported "
+                f"types: {list(TYPE_TO_ENUM.keys())}"
+            )
 
-        predictions_df = self._predict_pandas(data_df, **kwargs)
-        return convert_pandas_to_batch_type(
-            predictions_df,
-            type=TYPE_TO_ENUM[type(data)],
-            cast_tensor_columns=self._cast_tensor_columns,
-        )
+        has_predict_numpy = self.__class__._predict_numpy != Predictor._predict_numpy
+        has_predict_pandas = self.__class__._predict_pandas != Predictor._predict_pandas
+        if not has_predict_numpy and not has_predict_pandas:
+            raise NotImplementedError(
+                "None of `_predict_pandas` or `_predict_numpy` are "
+                f"implemented for input data batch format `{batch_format}`."
+            )
```

Comment:
Ah I see that the check is here, nice. This happens upstream of any `Predictor._batch_format_to_use()` calls, right?

Reply:
this one is a bit more downstream tho upon seeing data on `predict`, so i've added your suggestion above to surface the issue earlier